### ISSUE TITLE
Change the type of length from c_long to isize (in PySlice::index)

### ISSUE THREAD

**Issue by cmpute** [PyO3/pyo3#3756]
https://github.com/PyO3/pyo3/blob/3af73fabcfda5ecd1357819bb8f475bff2f4f467/src/types/slice.rs#L69

The underlying function `PySlice_GetIndicesEx` has an input `length` of type `Py_ssize_t`, which should be mapped to `isize`, instead of `c_long`. The conversion from `c_long` to `Py_ssize_t` seems unnecessary.

**Comment by davidhewitt** [comment]
Agreed, a PR to fix this would be greatly appreciated 👍

**Comment by davidhewitt** [comment]
That said, I assume the length should be non-negative? Maybe `usize` is better, and we can overflow if the `usize` is greater than `isize::MAX`?

**Comment by cmpute** [comment]
Agree!. And if more changes are allowed, I propose further change the type of start/stop/length of `PySliceIndices` all to `usize`, and implement use the recommended new APIs `PySlice_Unpack` and `PySlice_AdjustIndices`. (according to [official C-API docs](https://docs.python.org/3/c-api/slice.html#c.PySlice_GetIndicesEx))

**Comment by davidhewitt** [comment]
I'm certainly open to making those changes to fit better with what Python offers, but we should consider the compatibility story. There's a lot changing in 0.21 already and I'm cautious of adding too much more.

**Comment by cmpute** [comment]
> I'm certainly open to making those changes to fit better with what Python offers, but we should consider the compatibility story. There's a lot changing in 0.21 already and I'm cautious of adding too much more.

This change is already a break change, but I will try to make the changes small, when I create the PR probably this or next week.
